streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -209,6 +211,10 @@ boolean setState(final State newState) {
             stateListener.onChange(this, state, oldState);
         }
 
+        if (oldState == State.CREATED && newState == State.PENDING_SHUTDOWN) {
+            completeShutdown(true);
```

Comment:
I'm wondering if we can avoid injecting the thread logic in the `setState` function. One idea is to not allow the transition from `CREATED` to `PENDING_SHUTDOWN`, and hence the `setState(PENDING_SHUTDOWN)` in `KafkaStreams#shutdown()` will return false (note all other states except `DEAD` can transit to `PENDING_SHUTDOWN`), and then we can depend on the return value i.e. in `shutdown()`:

```
if (!setState(PENDING_SHUTDOWN) && state() == CREATED)) {
    completeShutdown(true);
}
```

Reply:
If we did that we could still race with with the StreamThread itself and do stuff in run() while the caller of shutdown() is in completeShutdown(). I do agree that putting this logic in setState is ugly. Another option would be to have a casState (compare and swap state) function so the caller can be sure which state transition its making.

Reply:
But couldn't we have a lock on the state within `completeShutdown()` ? Thus, if we are not running yet, and enter `completeShutdown()` we also don't allow to transit into RUNNING -- and if we did leave CREATED, we "cancel" `completeShutdown()` ?